File: packages/android-client/android/src/main/java/voltra/widget/server/WidgetServerRequestBuilder.kt
package voltra.widget.server

import android.content.Context
import android.content.res.Configuration
import android.net.Uri
import android.os.Build
import android.util.Log
import java.net.HttpURLConnection
import java.net.URL

/**
 * A request, fully decided, before anything opens a socket. Keeping this separate from the
 * connection is what lets the request contract be unit-tested: what a backend sees is a value, not
 * a side effect.
 */
data class WidgetServerRequest(
    val url: URL,
    val method: String,
    val headers: Map<String, String>,
    val body: ByteArray?,
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is WidgetServerRequest) return false

        return url == other.url &&
            method == other.method &&
            headers == other.headers &&
            body.contentEqualsOrBothNull(other.body)
    }

    override fun hashCode(): Int {
        var result = url.hashCode()
        result = 31 * result + method.hashCode()
        result = 31 * result + headers.hashCode()
        result = 31 * result + (body?.contentHashCode() ?: 0)
        return result
    }
}

private fun ByteArray?.contentEqualsOrBothNull(other: ByteArray?): Boolean =
    if (this == null || other == null) this == null && other == null else contentEquals(other)

/**
 * Turns resolved settings plus Voltra's own request parameters into the request the device sends.
 *
 * Both engines build their requests here, so a payload widget and a Dynamic Widget send the same
 * shape and the app's runtime overrides apply to both. Only the response and what the device does
 * with it differ.
 */
object WidgetServerRequestBuilder {
    private const val TAG = "VoltraWidgetServerReq"

    private const val CONNECT_TIMEOUT_MS = 15_000
    private const val READ_TIMEOUT_MS = 15_000

    /**
     * @param etag from the last `200`, sent as `If-None-Match`. Callers pass null when the stored
     *   ETag belongs to a different URL than the one being fetched now.
     * @return null when there is nothing to fetch — no URL, or fetching is off.
     */
    fun build(
        context: Context,
        scope: WidgetScope,
        settings: ResolvedWidgetServerSettings,
        etag: String? = null,
    ): WidgetServerRequest? {
        if (!settings.shouldFetch) {
            return null
        }

        val method = settings.method.uppercase()

        val builder =
            Uri
                .parse(settings.url)
                .buildUpon()
                .appendQueryParameter("widgetId", scope.widgetId)
                .appendQueryParameter("platform", "android")
                .appendQueryParameter("theme", currentTheme(context))
                .appendQueryParameter("locale", currentLocale(context))

        // Voltra's own keys are appended first and the app's keys are rejected at call time if they
        // collide, so nothing here can shadow what the server relies on.
        settings.query.forEach { (key, value) -> builder.appendQueryParameter(key, value) }

        val headers = mutableMapOf<String, String>()
        headers["Accept"] = "application/json"
        headers["User-Agent"] = userAgent()
        headers.putAll(settings.headers)

        if (etag != null) {
            headers["If-None-Match"] = etag
        }

        var body = settings.body?.toByteArray(Charsets.UTF_8)

        if (body != null && method in WidgetServerUpdateDefaults.BODYLESS_METHODS) {
            // HttpURLConnection silently turns a GET with an output stream into a POST, which would
            // hit a different endpoint than the app asked for. Dropping the body is the lesser
            // surprise, and it is documented.
            Log.w(TAG, "Dropping request body for widget '${scope.widgetId}': $method cannot carry one")
            body = null
        }

        if (body != null) {
            headers["Content-Type"] = "application/json"
        }

        return WidgetServerRequest(
            url = URL(builder.build().toString()),
            method = method,
            headers = headers,
            body = body,
        )
    }

    /** Opens and configures a connection for [request]. The caller connects, reads, and disconnects. */
    fun open(request: WidgetServerRequest): HttpURLConnection {
        val connection = request.url.openConnection() as HttpURLConnection

        connection.requestMethod = request.method
        connection.connectTimeout = CONNECT_TIMEOUT_MS
        connection.readTimeout = READ_TIMEOUT_MS
        // Redirects are followed by WidgetServerFetcher instead, which refuses to leave the host
        // the app configured so an Authorization header cannot be replayed somewhere else.
        connection.instanceFollowRedirects = false

        request.headers.forEach { (key, value) -> connection.setRequestProperty(key, value) }

        request.body?.let { body ->
            connection.doOutput = true
            connection.outputStream.use { it.write(body) }
        }

        return connection
    }

    fun currentTheme(context: Context): String {
        val nightModeFlags = context.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
        return if (nightModeFlags == Configuration.UI_MODE_NIGHT_YES) "dark" else "light"
    }

    fun currentLocale(context: Context): String {
        val configuration = context.resources.configuration
        val locale =
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                configuration.locales[0]
            } else {
                @Suppress("DEPRECATION")
                configuration.locale
            }

        return locale?.toLanguageTag() ?: "en"
    }

    private fun userAgent(): String =
        "VoltraWidget/${voltra.BuildConfig.VOLTRA_VERSION} (Android/${Build.VERSION.RELEASE})"
}

File: packages/ios-client/ios/shared/WidgetServer/WidgetServerRequestBuilder.swift
import Foundation

/// The device state a request carries that is not part of the settings: the theme and locale the
/// widget is being drawn in, and who is asking.
///
/// Passed in rather than read here so the request contract can be tested without UIKit, and so a
/// timeline request built for one appearance cannot accidentally report another.
public struct WidgetServerRequestContext: Equatable, Sendable {
  public let theme: String
  public let locale: String
  public let userAgent: String
  /// Only payload widgets send `family`: one Dynamic Widget fetch serves every size, so its props
  /// must be size-agnostic and the entry picks its layout from `env.widgetFamily`.
  public let family: String?

  public init(theme: String, locale: String, userAgent: String, family: String? = nil) {
    self.theme = theme
    self.locale = locale
    self.userAgent = userAgent
    self.family = family
  }
}

/// Turns resolved settings plus Voltra's own request parameters into the request the device sends.
///
/// Both engines build their requests here, so a payload widget and a Dynamic Widget send the same
/// shape and the app's runtime overrides apply to both. Only the response and what the device does
/// with it differ.
public enum WidgetServerRequestBuilder {
  /// Widgets have limited execution time, so a request that has not answered in this long is not
  /// going to be useful even if it eventually does.
  public static let timeoutSeconds: TimeInterval = 15

  /// - Parameter etag: from the last `200`, sent as `If-None-Match`. Callers pass nil when the
  ///   stored ETag belongs to a different URL than the one being fetched now.
  /// - Returns: nil when there is nothing to fetch — no URL, fetching is off, or the URL will not
  ///   parse.
  public static func build(
    scope: WidgetScope,
    settings: ResolvedWidgetServerSettings,
    context: WidgetServerRequestContext,
    etag: String? = nil,
    // The placement's merged configuration (ADR 0007). Empty for a widget with no configuration
    // parameters, or when the caller only knows the scope's key — the caller is responsible for
    // passing the same map that produced `scope`'s key, since the key alone cannot be reversed
    // back into the configuration that produced it.
    configuration: [String: String] = [:]
  ) -> URLRequest? {
    guard settings.shouldFetch, let url = settings.url, var components = URLComponents(string: url) else {
      return nil
    }

    var queryItems = components.queryItems ?? []
    queryItems.append(URLQueryItem(name: "widgetId", value: scope.widgetId))
    queryItems.append(URLQueryItem(name: "platform", value: "ios"))

    if let family = context.family {
      queryItems.append(URLQueryItem(name: "family", value: family))
    }

    queryItems.append(URLQueryItem(name: "theme", value: context.theme))
    queryItems.append(URLQueryItem(name: "locale", value: context.locale))

    // Present for every method, absent when the widget has no configuration parameters (ADR
    // 0007). `instance` is the hash of the canonical `configuration`, sent so a backend can cache
    // or log per instance without recomputing it.
    if case let .instance(_, key) = scope, let canonicalConfiguration = WidgetCanonicalConfiguration.canonicalize(configuration) {
      queryItems.append(URLQueryItem(name: "instance", value: key))
      queryItems.append(URLQueryItem(name: "configuration", value: canonicalConfiguration))
    }

    // Voltra's own keys are appended first and the app's keys are rejected at call time if they
    // collide, so nothing here can shadow what the server relies on.
    for key in settings.query.keys.sorted() {
      queryItems.append(URLQueryItem(name: key, value: settings.query[key]))
    }

    components.queryItems = queryItems

    guard let resolvedUrl = components.url else { return nil }

    var request = URLRequest(url: resolvedUrl)
    let method = settings.method.uppercased()

    request.httpMethod = method
    request.timeoutInterval = timeoutSeconds
    request.setValue("application/json", forHTTPHeaderField: "Accept")
    request.setValue(context.userAgent, forHTTPHeaderField: "User-Agent")

    for (key, value) in settings.headers {
      request.setValue(value, forHTTPHeaderField: key)
    }

    if let etag {
      request.setValue(etag, forHTTPHeaderField: "If-None-Match")
    }

    if let body = settings.body {
      if WidgetServerUpdateDefaults.bodylessMethods.contains(method) {
        // URLSession sends a Content-Length and drops the body on GET, so the server would see a
        // request the app did not mean to send. Dropping it here is the lesser surprise, and it is
        // documented.
        VoltraLogger.widget.warning(
          "Dropping request body for widget '\(scope.widgetId, privacy: .public)': \(method, privacy: .public) cannot carry one"
        )
      } else {
        request.httpBody = Data(body.utf8)
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
      }
    }

    return request
  }
}
